src/pci/src/configuration.rs

```diff
@@ -698,108 +649,95 @@ impl PciConfiguration {
+
+        self.registers[reg_idx + 1] = (addr >> 32) as u32;
+        self.writable_bits[reg_idx + 1] = 0xffff_ffff;
+        self.bars[bar_idx + 1].addr = self.registers[reg_idx + 1];
+        self.bars[bar_idx].size = bar_size_lo;
+        self.bars[bar_idx + 1].size = bar_size_hi;
+        self.bars[bar_idx + 1].used = true;
+
+        // Addresses of memory BARs are 16-byte aligned so the lower 4 bits are always 0. Within
+        // the register we use this 4 bits to encode extra information about the BAR. The meaning
+        // of these bits is:
+        //
+        // |    Bit 3     | Bits 2-1 |  Bit 0   |
+        // | Prefetchable |   type   | Always 0 |
+        self.registers[reg_idx] = (((addr & 0xffff_ffff) as u32) & BAR_MEM_ADDR_MASK)
+            | (PciBarPrefetchable::NotPrefetchable as u32)
+            | (PciBarRegionType::Memory64BitRegion as u32);
+        self.writable_bits[reg_idx] = BAR_MEM_ADDR_MASK;
         self.bars[bar_idx].addr = self.registers[reg_idx];
         self.bars[bar_idx].used = true;
-        self.bars[bar_idx].r#type = Some(config.region_type);
 
         Ok(())
     }
 
     /// Returns the address of the given BAR region.
-    pub fn get_bar_addr(&self, bar_num: usize) -> u64 {
-        let bar_idx = BAR0_REG + bar_num;
-
-        let mut addr = u64::from(self.bars[bar_num].addr & self.writable_bits[bar_idx]);
+    ///
+    /// This assumes that `bar_idx` is a valid BAR register.
+    pub fn get_bar_addr(&self, bar_idx: usize) -> u64 {
+        assert!(bar_idx < NUM_BAR_REGS);
 
-        if let Some(bar_type) = self.bars[bar_num].r#type {
-            if bar_type == PciBarRegionType::Memory64BitRegion {
-                addr |= u64::from(self.bars[bar_num + 1].addr) << 32;
-            }
-        }
+        let reg_idx = BAR0_REG + bar_idx;
 
-        addr
+        (u64::from(self.bars[bar_idx].addr & self.writable_bits[reg_idx]))
+            | (u64::from(self.bars[bar_idx + 1].addr) << 32)
     }
 
     /// Adds the capability `cap_data` to the list of capabilities.
     ///
     /// `cap_data` should not include the two-byte PCI capability header (type, next).
     /// Correct values will be generated automatically based on `cap_data.id()` and
     /// `cap_data.len()`.
-    pub fn add_capability(&mut self, cap_data: &dyn PciCapability) -> Result<usize> {
+    pub fn add_capability(&mut self, cap_data: &dyn PciCapability) -> Result<usize, PciError> {
         let total_len = cap_data.bytes().len() + 2;
         let (cap_offset, tail_offset) = match self.last_capability {
             Some((offset, len)) => (Self::next_dword(offset, len), offset + 1),
             None => (FIRST_CAPABILITY_OFFSET, CAPABILITY_LIST_HEAD_OFFSET),
         };
         let end_offset = cap_offset
             .checked_add(total_len)
-            .ok_or(Error::CapabilitySpaceFull(total_len))?;
+            .ok_or(PciError::CapabilitySpaceOverflow(total_len, cap_offset))?;
```

Comment:
Same question about asserts.

Reply:
same here. Turned these into assertions and updated tests.